### ISSUE TITLE
The attention mechanism should always be applied

### ISSUE THREAD

**Issue by bdvllrs** [ruflab/shimmer#27]
We are considering the attention mechanism as separate, ad-hoc module that would only affect GW encoding in special cases. This complements the discussion in #19.
After a discussion with @rufinv, we found that the attention mechanism should always be here.

The general idea is: 
- it sits after encoding the unimodal representation and before encoding them into pre-fusion GW representations
- it takes the encoded unimodal representations as input, and returns coefficients for each of them
- these coefficients are given as parameter to `GWModule.encode` that will multiply the representations (if a domain is absent, it automatically gains the 0 coefficient)
- Loss modules have access to original representations and the attention coefficients

In this way, @RolandBERTINJOHANNET's fusion mechanism coefficients (90% of 1/0 for domains and 10% of a mix) would not be hardcoded but given by a special Attention module.

Some Attention modules would be:
- BinaryAttention (1 if domain is here, 0 otherwise) this what is used for the GW with cycle losses
- RandomAttention (1/0 with given probability, random otherwise) this is what is used by @RolandBERTINJOHANNET in his fusion
- Other Attention methods: this is to be defined by @larascipio or @HugoChateauLaurent.

In terms of implementation, I think an Attention module would just be a `torch.nn.Module` where forwards returns the coefficients for each domain:
```python
class AttentionBase(torch.nn.Module, ABC):
    @abstractmethod
    def forward(self, domains: LatentsDomainGroupT) -> dict[str, float]: ...
```

and this module would be an input to GlobalWorkspaceBase class.
